crates/findex/src/pyo3/py_api.rs

```diff
@@ -536,3 +518,213 @@ impl InternalFindex {
+        let mut token = pyo3_unwrap!(Token::from_str(&token), "error reading token");
+
+        pyo3_unwrap!(
+            token.reduce_permissions(search, index),
+            "error reducing token's permissions"
+        );
+
+        Ok(token.to_string())
+    }
+
+    /// Generate a new random Findex Cloud token
+    #[staticmethod]
+    pub fn generate_new_token(
+        index_id: String,
+        fetch_entries_seed: &[u8],
+        fetch_chains_seed: &[u8],
+        upsert_entries_seed: &[u8],
+        insert_chains_seed: &[u8],
+    ) -> PyResult<String> {
+        let token = pyo3_unwrap!(
+            Token::random_findex_master_key(
+                index_id,
+                pyo3_unwrap!(
+                    uint8slice_to_seed(fetch_entries_seed, "fetch_entries_seed"),
+                    "error reading seed from Uint8Slice"
+                ),
+                pyo3_unwrap!(
+                    uint8slice_to_seed(fetch_chains_seed, "fetch_chains_seed"),
+                    "error reading seed from Uint8Slice"
+                ),
+                pyo3_unwrap!(
+                    uint8slice_to_seed(upsert_entries_seed, "upsert_entries_seed"),
+                    "error reading seed from Uint8Slice"
+                ),
+                pyo3_unwrap!(
+                    uint8slice_to_seed(insert_chains_seed, "insert_chains_seed"),
+                    "error reading seed from Uint8Slice"
+                ),
+            ),
+            "error generating new token"
+        );
+
+        Ok(token.to_string())
+    }
+}
+
+/// Conversion helpers from Python types to Rust
+
+fn uint8slice_to_seed(
+    seed: &[u8],
+    debug_name: &str,
+) -> Result<KeyingMaterial<SIGNATURE_SEED_LENGTH>, FindexErr> {
+    KeyingMaterial::try_from_bytes(seed).map_err(|_| {
+        FindexErr::Other(format!(
+            "{debug_name} is of wrong size ({SIGNATURE_SEED_LENGTH} expected)",
+        ))
+    })
+}
+
+fn indexed_values_and_keywords_to_rust(
+    indexed_values_and_strings: HashMap<IndexedValuePy, Vec<&str>>,
+) -> HashMap<IndexedValueRust, HashSet<Keyword>> {
+    let mut indexed_values_and_keywords = HashMap::with_capacity(indexed_values_and_strings.len());
+    for (indexed_value, strings) in indexed_values_and_strings {
+        let mut keywords = HashSet::with_capacity(strings.len());
+        for string in strings {
+            keywords.insert(Keyword::from(string));
+        }
+        indexed_values_and_keywords.insert(indexed_value.0, keywords);
+    }
+    indexed_values_and_keywords
+}
+
+fn search_results_to_python(
+    search_results: HashMap<Keyword, HashSet<Location>>,
+    py: Python,
+) -> HashMap<String, Vec<PyObject>> {
+    search_results
+        .iter()
+        .map(|(keyword, locations)| {
+            (
+                // Convert keyword to string or base64
+                match keyword.clone().try_into_string() {
+                    Ok(s) => s,
+                    Err(_) => format!("{keyword:?}"),
```

Comment:
I think the debug does not convert to base64 anymore, you should convert manually here.